### PR TITLE
Make cargohold redirect target configurable by requesting host

### PR DESCRIPTION
**Pull request by supersven**
This is needed for multi-ingress support, where we want to redirect an app to a S3 store download URL according to the sent `Host` header. (Multi-ingress means that there are separate domains for the same backend to obfuscate relationships between clients and the backend.)

This sequence diagram from the documentation (which is also a good entry point to this PR) shows how this should work:
image

The longer story is told here: [Confluence: Multi-Ingress (Proxy Mode V2)](https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/813137982/Multi-Ingress+Proxy+Mode+V2).

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: integration/test/Test/AssetDownload.hs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        resp.status `shouldMatchInt` 404
+      bindResponse (downloadAsset' user key "red.example.com" noRedirects) $ \resp -> do
+        resp.status `shouldMatchInt` 302
+        locationHeaderHost resp `shouldMatch` "s3-download.red.example.com"
+      bindResponse (downloadAsset' user key "green.example.com" noRedirects) $ \resp -> do
+        resp.status `shouldMatchInt` 302
+        locationHeaderHost resp `shouldMatch` "s3-download.green.example.com"
+      bindResponse (downloadAsset' user key "unknown.example.com" noRedirects) $ \resp -> do
+        resp.status `shouldMatchInt` 404
+        resp.json %. "label" `shouldMatch` "no-asset-endpoint"
```
Comment (fisx) [inline integration/test/Test/AssetDownload.hs:49]:
```suggestion
        resp.json %. "label" `shouldMatch` "not-found"
```

also test label in the block in line 40 above.  this way you don't leak the fact that the server has a multi-ingress setup.

File: integration/test/Test/AssetDownload.hs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      setField "aws.s3DownloadEndpoint" "http://s3-download.example.com" v
+        >>= setField "aws.multiIngress" multiIngressConfig
+
+    multiIngressConfig =
+      object
+        [ "red.example.com" .= "http://s3-download.red.example.com",
+          "green.example.com" .= "http://s3-download.green.example.com"
+        ]
+
+    locationHeaderHost :: Response -> String
```
Comment (fisx) [inline integration/test/Test/AssetDownload.hs:65]: Move this to Testlib, together with `locationHeader` as a separate helper?

**Review by fisx — COMMENTED** [review]
done reading the tests, looking good so far!

File: services/cargohold/src/CargoHold/S3.hs [hunk 3]
```diff
@@ -221,12 +242,46 @@ signedURL path = do
     toUri x = case parseURI strictURIParserOptions x of
       Left e -> do
         Log.err $
-          "remote" .= val "S3"
+          "remote"
+            .= val "S3"
             ~~ msg (val "Failed to generate a signed URI")
             ~~ msg (show e)
```
Comment (fisx) [inline services/cargohold/src/CargoHold/S3.hs:248]: interesting.  i didn't know `msg` was self-concatenating.

**Review by fisx — APPROVED** [review]

File: services/cargohold/src/CargoHold/API/Public.hs [hunk 4]
```diff
@@ -146,24 +146,26 @@ downloadAssetV3 ::
   AssetKey ->
   Maybe AssetToken ->
   Maybe AssetToken ->
+  Maybe Text ->
   Handler (Maybe (AssetLocation Absolute))
-downloadAssetV3 usr key tok1 tok2 = do
-  AssetLocation <$$> V3.download (mkPrincipal usr) key (tok1 <|> tok2)
+downloadAssetV3 usr key tok1 tok2 mbHostHeader = do
```
[the remark this answers is not in the document]
Comment (supersven) [reply services/cargohold/src/CargoHold/API/Public.hs:151]: Well, my argumentation goes like this:
- For customers with an old configuration, this endpoint does not change.
- If I would move this change to `V4`, then we would also have to get the app developers to adjust and release the app (which may take months :roll_eyes: )

File: libs/wire-api/src/Wire/API/Routes/Public.hs [hunk 5]
```diff
@@ -31,7 +31,7 @@ module Wire.API.Routes.Public
     ZConversation,
     ZProvider,
     DescriptionOAuthScope,
-    ZHost,
+    ZHostOpt,
```
[the remark this answers is not in the document]
Comment (supersven) [reply libs/wire-api/src/Wire/API/Routes/Public.hs:34]: @battermann thought he needed the `Z-Host` header for some OAuth related changes. Turned out that he didn't, but he was so kind to leave the necessary changes in for me.

File: docs/src/developer/reference/config-options.md [hunk 6]
```diff
@@ -722,3 +722,67 @@ config.disabledAPIVersions: [ v3 ]
 ```
 
 The default setting is that no API version is disabled.
+
+## Settings in cargohold
+
+### (Fake) AWS
+
+AWS S3 (or an alternative provider / service) is used to upload and download
+assets. The Haddock of `AWSOpts` provides a lot of useful information.
```
[the remark this answers is not in the document]
Comment (supersven) [reply docs/src/developer/reference/config-options.md:731]: Unfortunately, I don't know such a stable way either. That's why I initially provided no link. But, linking to the source code seems to be a better idea. :+1:

File: integration/test/Test/AssetDownload.hs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+  key <- bindResponse (uploadAsset user) $ \resp -> do
+    resp.status `shouldMatchInt` 201
+    resp.json %. "key"
+
+  void . bindResponse (downloadAsset user key id) $ \resp -> do
+    resp.status `shouldMatchInt` 200
+    assertBool
+      ("Expect 'Hello World!' as text asset content. Got: " ++ show resp.body)
+      (resp.body == fromString "Hello World!")
```
[the remark this answers is not in the document]
Comment (supersven) [reply integration/test/Test/AssetDownload.hs:25]: The multipart contents of the [`/assets` endpoint](https://staging-nginz-https.zinfra.io/v4/api/swagger-ui/#/default/post_assets) are special: The first part contains meta-data, the second part the file content. The file is downloaded from the assets host (a fake S3 service); thus, it only contains the content part.

File: integration/test/Test/AssetDownload.hs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+testDownloadAssetMultiIngressS3DownloadUrl :: HasCallStack => App ()
+testDownloadAssetMultiIngressS3DownloadUrl = do
+  user <- randomUser OwnDomain def
+
+  key <- bindResponse (uploadAsset user) $ \resp -> do
+    resp.status `shouldMatchInt` 201
+    resp.json %. "key"
```
[the remark this answers is not in the document]
Comment (supersven) [reply integration/test/Test/AssetDownload.hs:33]: I've extracted the *download and assert* part into a test helper function and now cover both ways. Thanks for bringing this up :+1:

File: services/cargohold/src/CargoHold/S3.hs [hunk 9]
```diff
@@ -101,12 +101,18 @@ uploadV3 ::
...
+      ~~ "asset.owner"
+        .= toByteString prc
+      ~~ "asset.key"
+        .= key
+      ~~ "asset.type_from_request_ignored"
+        .= MIME.showType (V3.hdrType originalHeaders)
+      ~~ "asset.type"
+        .= MIME.showType ct
+      ~~ "asset.size"
+        .= cl
```
[the remark this answers is not in the document]
Comment (supersven) [reply services/cargohold/src/CargoHold/S3.hs]: I don't know what went wrong (maybe nervous fingers :roll_eyes: ). I've cleaned it up and the formatting seems to be stable.

File: services/cargohold/src/CargoHold/S3.hs (hunk 3 above)
Comment (supersven) [reply services/cargohold/src/CargoHold/S3.hs]: It actually is: https://hoogle.zinfra.io/file/nix/store/m3hbxxbjk8j6a13gym6vhswdqx2z8a89-tinylog-0.15.0-doc/share/doc/tinylog-0.15.0/html/src/System.Logger.Message.html#msg. But, I have to admit I didn't knew it either :smile_cat:

File: integration/test/Test/AssetDownload.hs (hunk 1 above)
Comment (supersven) [reply integration/test/Test/AssetDownload.hs]: Yep, that's a good idea :+1:

File: docs/src/developer/reference/config-options.md [hunk 10]
```diff
@@ -722,3 +722,67 @@ config.disabledAPIVersions: [ v3 ]
...
+The backend domain of a download request is defined by its `Z-Host` header which
+is set by `nginz`. (Multi-ingress handlling only applies to download requests as
+these are implemented by redirects to the S3 assets host for local assets.
+Uploads are handled by cargohold directly itself.)
+
+The config `aws.multiIngress` is a map from backend domain (`Z-Host` header
+value) to a S3 download endpoint. The `Z-Host` header is set by `nginz` to the
+value of the incoming requests `Host` header. If there's no config map entry for
+a provided `Z-Host` in a download request for a local asset, then an error is
+returned.
```
[the remark this answers is not in the document]
Comment (supersven) [reply docs/src/developer/reference/config-options.md:760]: Added two comments to tickets:
- https://wearezeta.atlassian.net/browse/WPB-1142?focusedCommentId=113816
- https://wearezeta.atlassian.net/browse/WPB-1138?focusedCommentId=113818

As discussed, let's get this PR merged to make testing easier.

**Pull request merged**